### PR TITLE
Add a real-er web restart, doctor, workflow

### LINKED ISSUE #33767: Weird unexplained bump in consumer dependency count
https://github.com/flutter/flutter/pull/33533 increased the consumer dependency count somehow.

cc @jonahwilliams

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Adds an implementation of hot restart for that web that refreshes the page. Wires up the web hot runner to use the webkit inspection protocol to refresh, close, and listen to page events.

Refactors some of the web workflow APis and adds tests.

Adds devicelab tests (expected to fail until we install chrome) to test this workflow end to end.

------
Unrelated Fixes https://github.com/flutter/flutter/issues/33767

### CODE REVIEW

File: packages/flutter_tools/lib/src/resident_web_runner.dart [hunk 1]
```diff
@@ -58,12 +62,16 @@ class ResidentWebRunner extends ResidentRunner {
   }
 
   @override
-  Future<void> cleanupAfterSignal() {
+  Future<void> cleanupAfterSignal() async {
+    await _connection.sendCommand('Browser.close');
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/resident_web_runner.dart]: nit: do we usually capitalize command names?

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      final String path = fs.path.join(prefix, kWindowsExecutable);
+      return fs.file(path).existsSync();
+    }, orElse: () => '.');
+    return fs.path.join(windowsPrefix, kWindowsExecutable);
+  }
+  throwToolExit('Platform ${platform.operatingSystem} is not supported.');
+  return null;
+}
+
+// Responsible for launching chrome with devtools configured.
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/web/chrome.dart]: nit: `///`

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      _currentCompleter = Completer<Chrome>();
+    }
+    chromeConnection.close();
+    _process?.kill(ProcessSignal.SIGKILL);
+    await _process?.exitCode;
+    try {
+      // Chrome starts another process as soon as it dies that modifies the
+      // profile information. Give it some time before attempting to delete
+      // the directory.
+      await Future<void>.delayed(const Duration(milliseconds: 500));
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/web/chrome.dart]: Should this be Windows-only?

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/resident_web_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_web_runner.dart]: This is a webkit inspection protocol command

File: packages/flutter_tools/lib/src/web/chrome.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/web/chrome.dart]: I don't know...

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+      await _dataDir?.delete(recursive: true);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/web/chrome.dart]: Move to finally {}

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    await _process?.exitCode;
+    try {
+      // Chrome starts another process as soon as it dies that modifies the
+      // profile information. Give it some time before attempting to delete
+      // the directory.
+      await Future<void>.delayed(const Duration(milliseconds: 500));
+    } catch (_) {
+      // Silently fail if we can't clean up the profile information.
+    } finally {
+      await _dataDir?.delete(recursive: true);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/web/chrome.dart]: Probably also need to not die if this throws a FileSystemException if the delete fails.

**Comment by jonahwilliams** [comment]
Verified that this change succeeds to build in fuchsia-source

**Comment by cbracken** [comment]
This broke the Fuchsia roll. I've tried to revert, but there are merge conflicts. @jonahwilliams can you work on a revert.

**Pull request merged**